packages/cli/src/error_handler.ts

```diff
@@ -0,0 +1,81 @@
+import { COLOR, Printer } from '@aws-amplify/cli-core';
+import { InvalidCredentialError } from './error/credential_error.js';
+import { EOL } from 'os';
+import { Argv } from 'yargs';
+
+/**
+ * Attaches process listeners to handle unhandled exceptions and rejections
+ */
+export const attachUnhandledExceptionListeners = (): void => {
+  process.on('unhandledRejection', (reason) => {
+    process.exitCode = 1;
+    if (reason instanceof Error) {
+      handleError(reason);
+    } else if (typeof reason === 'string') {
+      handleError(new Error(reason));
+    } else {
+      // This should never happen unless something truly strange has happened
+      Printer.print(`Cannot handle rejection of type [${typeof reason}]`);
+    }
+  });
+
+  process.on('uncaughtException', (error) => {
+    process.exitCode = 1;
+    handleError(error);
+  });
+};
+
+/**
+ * Generates a function that is intended to be used as a callback to yargs.fail()
+ * All logic for actually handling errors should be delegated to handleError.
+ *
+ * For some reason the yargs object that is injected into the fail callback does not include all methods on the Argv type
+ * This generator allows us to inject the yargs parser into the callback so that we can call parser.exit() from the failure handler
+ * This prevents our top-level error handler from being invoked after the yargs error handler has already been invoked
+ */
+export const generateCommandFailureHandler = (
+  parser: Argv
+): ((message: string, error: Error) => void) => {
+  /**
+   * Format error output when a command fails
+   * @param message error message set by the yargs:check validations
+   * @param error error thrown by yargs handler
+   */
+  const handleCommandFailure = (message: string, error: Error) => {
+    const printHelp = () => {
+      Printer.printNewLine();
+      parser.showHelp();
+      Printer.printNewLine();
+    };
+
+    handleError(error, printHelp, message);
+    parser.exit(1, error);
+  };
+  return handleCommandFailure;
+};
+
+/**
+ * Root level error handling for uncaught errors during CLI command execution
```

Comment:
nit: if this is also used in `generateCommandFailureHandler` then `Root level` might not be 100% accurate.

Reply:
Updated the comment.